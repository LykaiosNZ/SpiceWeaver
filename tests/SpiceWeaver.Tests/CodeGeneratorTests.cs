namespace SpiceWeaver.Tests;

[TestFixture(TestOf = typeof(CodeGenerator))]
public sealed class CodeGeneratorTests
{
    [Test]
    [TestCaseSource(nameof(NullOrWhitespace))]
    public void GenerateFromJson_WhenSchemaJsonIsNullOrWhitespace_ShouldThrow(string? json)
    {
        Assert.Throws<ArgumentException>(() => CodeGenerator.Generate("foo", "bar", json!));
    }

    [Test]
    [TestCaseSource(nameof(NullOrWhitespace))]
    public void GenerateFromJson_WhenNamespaceIsNullOrWhitespace_ShouldThrow(string? @namespace)
    {
        Assert.Throws<ArgumentException>(() => CodeGenerator.Generate(@namespace!, "bar", TestSchema.Json));
    }
    
    [Test]
    [TestCaseSource(nameof(NullOrWhitespace))]
    public void GenerateFromJson_WhenClassNameIsNullOrWhitespace_ShouldThrow(string? className)
    {
        Assert.Throws<ArgumentException>(() => CodeGenerator.Generate("foo", className!, TestSchema.Json));
    }

    [Test]
    public void GenerateFromJson_ShouldGenerateExpectedOutput()
    {
        var output = CodeGenerator.Generate("TestNamespace", "TestSchema", TestSchema.Json);
        
        Snapshot.Match(output);
    }

    [Test]
    [TestCaseSource(nameof(NullOrWhitespace))]
    public void GenerateFromSchema_WhenNamespaceIsNullOrWhitespace_ShouldThrow(string? @namespace)
    {
        Assert.Throws<ArgumentException>(() => CodeGenerator.Generate(@namespace!, "bar", TestSchema.Object));
    }
    
    [Test]
    [TestCaseSource(nameof(NullOrWhitespace))]
    public void GenerateFromSchema_WhenClassNameIsNullOrWhitespace_ShouldThrow(string? className)
    {
        Assert.Throws<ArgumentException>(() => CodeGenerator.Generate("foo", className!, TestSchema.Object));
    }

    [Test]
    public void GenerateFromSchema_ShouldGenerateExpectedOutput()
    {
        var output = CodeGenerator.Generate("TestNamespace", "TestSchema", TestSchema.Object);
        
        Snapshot.Match(output);
    }

    private static IEnumerable<string?> NullOrWhitespace =>
    [
        null,
        "",
        " ",
        "\r",
        "\t"
    ];
}