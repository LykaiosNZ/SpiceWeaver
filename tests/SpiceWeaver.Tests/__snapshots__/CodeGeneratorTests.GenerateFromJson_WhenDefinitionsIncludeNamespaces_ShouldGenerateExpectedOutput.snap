namespace TestNameSpace
{
    public static class TestSchema
    {
        public static class Definitions
        {
            public static class User
            {
                public const string Name = "user";
                public const string? Namespace = "mynamespace";
                public static string WithId(string id, bool includeNamespace = false) => includeNamespace ? $"mynamespace/user:{id}" : $"user:{id}";
            }

            public static class Document
            {
                public const string Name = "document";
                public const string? Namespace = "mynamespace";
                public static string WithId(string id, bool includeNamespace = false) => includeNamespace ? $"mynamespace/document:{id}" : $"document:{id}";
                public static class Relations
                {
                    public const string Viewer = "viewer";
                    public const string Editor = "editor";
                }

                public static class Permissions
                {
                    public const string View = "view";
                    public const string Edit = "edit";
                }
            }
        }
    }
}
